tests/integration/docarray_v2/test_v2.py

```diff
@@ -0,0 +1,142 @@
+from docarray.documents import Image
+from docarray.typing import AnyTensor, ImageUrl
+
+from jina import Deployment, Executor, Flow, requests
+
+
+def test_different_document_schema():
+    class Image(BaseDocument):
+        tensor: Optional[AnyTensor]
+        url: ImageUrl
+
+    class MyExec(Executor):
+        @requests(on='/foo')
+        def foo(self, docs: DocumentArray[Image], **kwargs) -> DocumentArray[Image]:
+            for doc in docs:
+                doc.tensor = doc.url.load()
+            return docs
+
+    with Flow().add(uses=MyExec) as f:
+        docs = f.post(
+            on='/foo',
+            inputs=DocumentArray[Image](
+                [Image(url='https://via.placeholder.com/150.png')]
+            ),
+            return_type=DocumentArray[Image],
+        )
+        docs = docs.stack()
+        assert docs.tensor.ndim == 4
+
+
+def test_send_custom_doc():
+    class MyDoc(BaseDocument):
+        text: str
+
+    class MyExec(Executor):
+        @requests(on='/foo')
+        def foo(self, docs: DocumentArray[MyDoc], **kwargs):
+            docs[0].text = 'hello world'
+
+    with Flow().add(uses=MyExec) as f:
+        doc = f.post(on='/foo', inputs=MyDoc(text='hello'))
+        assert doc[0].text == 'hello world'
+
+
+def test_input_output_type():
+    class MyDoc(BaseDocument):
+        text: str
+
+    class MyExec(Executor):
+        @requests(
+            on='/foo',
+            input_type=DocumentArray[MyDoc],
+            output_type=DocumentArray[MyDoc],
+        )
+        def foo(self, docs, **kwargs):
+            assert docs.__class__.document_type == MyDoc
+            docs[0].text = 'hello world'
+            return docs
+
+    with Flow().add(uses=MyExec) as f:
+        docs = f.post(
+            on='/foo', inputs=MyDoc(text='hello'), return_type=DocumentArray[MyDoc]
+        )
+        assert docs[0].text == 'hello world'
+        assert docs.__class__.document_type == MyDoc
+
+
+def test_input_output_type_annotation():
+    class MyDoc(BaseDocument):
+        text: str
+
+    class MyExec(Executor):
+        @requests(on='/bar')
+        def bar(self, docs: DocumentArray[MyDoc], **kwargs) -> DocumentArray[MyDoc]:
+            assert docs.__class__.document_type == MyDoc
+            docs[0].text = 'hello world'
+            return docs
+
+    with Flow().add(uses=MyExec) as f:
+        docs = f.post(
+            on='/bar', inputs=MyDoc(text='hello'), return_type=DocumentArray[MyDoc]
+        )
+        assert docs[0].text == 'hello world'
+        assert docs.__class__.document_type == MyDoc
+
+
+def test_different_output_input():
+    class InputDoc(BaseDocument):
+        img: Image
+
+    class OutputDoc(BaseDocument):
+        embedding: AnyTensor
+
+    class MyExec(Executor):
+        @requests(on='/bar')
+        def bar(
+            self, docs: DocumentArray[InputDoc], **kwargs
+        ) -> DocumentArray[OutputDoc]:
+            docs_return = DocumentArray[OutputDoc](
```

Comment:
can we also support implicit casting ?
I believe returning list of output doc or DocumentArray([OutputDoc()...]) is enough from a user perspective

Reply:
what do you mean by explicit casting? like we look at the return object and we use the class ? Gut feeling I don't like it but maybe ? I would stick to what FastAPI does tbh

Reply:
it looks like in fastapi you don't necessarily have to cast your response. https://fastapi.tiangolo.com/advanced/additional-responses/#additional-response-with-model
We already do one casting which is necessary and enough:
`[OutputDoc(embedding=np.zeros((100, 1))) for _ in range(len(docs))]`
The user doesn't really need to do again `DocumentArray[OutputDoc]` and if really needed, the runtime can do that for him

Reply:
but this is a v2 problem not a jina problem. Atm we don't infer the type of the DocumentArray based on the data so you need to precise which schema you DocumentArray will at da creation time.

We can discuss it in v2 development but IMO this is not a jina problem

Reply:
ah yeah my bad my initial response was wrong I misunderstood the question

Reply:
> Atm we don't infer the type of the DocumentArray based on the data
We don't have to infer the type from the data. We can infer from the type hint or the return type provided in the Executor definition, so it is a jina problem

Reply:
but when you create your document array you need to know what is the schema. How could we use the type hint from the function to do so ? I am not sure to understand

Reply:
lets handle this in the next PR if needed